First implementation of returning partial fragment on timeout

philiprodrigues opened the pull request:
This implements the first option described in #17 for dealing with low-rate inputs. (This PR looks like #19, which I mistakenly created against a branch that has already been merged - this PR is into develop)

philiprodrigues committed:
First implementation of returning partial fragment on timeout

philiprodrigues commented:
There's some behaviour in `DefaultRequestHandlerModel::check_waiting_requests()` that needs discussion: at run end, any waiting requests are immediately dropped (line 506 and beyond). With low-rate objects like trigger candidates, this means the last TC fragment is always empty. For this case, I think a better behaviour would be to just wait until the request times out and return the partial fragment. The timeout will happen eventually, so we won't wait indefinitely at run stop, so I think this is safe.

I think implementing my suggestion just means removing the block in `DefaultRequestHandlerModel::check_waiting_requests()` that begins `else if (!m_run_marker.load())`

roland-sipos commented on include/readoutlibs/concepts/RequestHandlerConcept.hpp:
```diff
@@ -55,13 +56,15 @@ class RequestHandlerConcept
   std::map<ResultCode, std::string> ResultCodeStrings{
     { ResultCode::kFound, "FOUND" },    { ResultCode::kNotFound, "NOT_FOUND" },
     { ResultCode::kTooOld, "TOO_OLD" }, { ResultCode::kNotYet, "NOT_YET_PRESENT" },
+    { ResultCode::kNotYet, "PARTIAL_RESULT" },
```
Should be `kPartial`, not `kNotYet`.

philiprodrigues committed:
At run stop, return what we have, not empty frag

Previous behaviour was that at run stop, pending data requests
returned an empty fragment. Now we wait until the number of requested
retries passes, and return what we have so far. This is for the
benefit of low-rate data objects, like trigger candidates

roland-sipos commented:
- Please find a mistake above with the enum.

- Now that the IOManager changes are on develop, could you please update the PR to reflect the changed function signatures? 

- Also, I would suggest that we test with the current changes, and we open another issue for your second comment: repond to kPartials on end of run, instead of dropping waiting requests on the floor:

> There's some behaviour in `DefaultRequestHandlerModel::check_waiting_requests()` that needs discussion: at run end, any waiting requests are immediately dropped (line 506 and beyond). With low-rate objects like trigger candidates, this means the last TC fragment is always empty. For this case, I think a better behaviour would be to just wait until the request times out and return the partial fragment. The timeout will happen eventually, so we won't wait indefinitely at run stop, so I think this is safe.
> 
> I think implementing my suggestion just means removing the block in `DefaultRequestHandlerModel::check_waiting_requests()` that begins `else if (!m_run_marker.load())`

Does this sound good for you?

philiprodrigues committed:
Merge branch 'develop' into philiprodrigues/return-partial-fragment-on-timeout to pick up iomanager-related changes

philiprodrigues committed:
Add TPSet topic option

philiprodrigues committed:
Fix mistake in enum declaration

roland-sipos reviewed (approved):
Went through it multiple times, looks good to me.
